synapse/handlers/room_member.py

```diff
@@ -1004,14 +1003,33 @@ def _remote_join(self, requester, remote_room_hosts, room_id, user, content):
         # filter ourselves out of remote_room_hosts: do_invite_join ignores it
         # and if it is the only entry we'd like to return a 404 rather than a
         # 500.
-
         remote_room_hosts = [
             host for host in remote_room_hosts if host != self.hs.hostname
         ]
 
         if len(remote_room_hosts) == 0:
             raise SynapseError(404, "No known servers")
 
+        if self.hs.config.limit_large_room_joins:
+            # Go fetch the room complexity here...
+            complexity_fetched = False
+            complexity = yield self.federation_handler.get_room_complexity(
+                remote_room_hosts, room_id
+            )
+
+            max_complexity = self.hs.config.limit_large_room_joins_complexity
+
+            if complexity:
+                if complexity["v1"] > max_complexity:
```

Comment:
I feel like this dereference should happen at a much lower level - probably in the federation_client. WDYT?

Reply:
Maybe? I don't know. We can probably change it later, if there's a better spot.